File: 2015/C/01/part_two.c
/******************************************************************************
 * Advent of Code 2015 day 01 part two
 *****************************************************************************/

#include <stdio.h> // fopen, fclose, fgetc, printf, fprintf
#include <stdlib.h> // exit

#define INPUT_FILE "../inputs/01-input.txt"

int main()
{
    int floor = 0;
    unsigned int index = 0;
    char ch = '\0';
    FILE *file = fopen(INPUT_FILE, "r");

    if (file == NULL)
    {
        fprintf(stderr, "Failed to open file '%s'\n", INPUT_FILE);
        exit(EXIT_FAILURE);
    }

    while ((ch = fgetc(file)) != EOF)
    {
        switch (ch)
        {
            case '(':
                floor++;
                break;

            case ')':
                floor--;
                break;

            default:
                fprintf(stderr, "Invalid character found at position '%u'\n", index + 1);
                fclose(file);
                exit(EXIT_FAILURE);
        }

        if (floor == -1)
        {
            printf("Basement is reached at position '%u'\n", index + 1);
            fclose(file);
            exit(EXIT_SUCCESS);
        }

        index++;
    }

    fprintf(stderr, "The basement was never reached?\n");
    fclose(file);
    exit(EXIT_FAILURE);
}

File: 2015/rust/01/part_two.rs
/*
 * Advent of Code 2015 day 01 part two
 */

use std::fs::File; // File::open
use std::io::Read; // read_to_string
use std::process; // process::exit

// String slice constant
const INPUT_FILE: &str = "../inputs/01-input.txt";

fn main() {
    let mut file = File::open(INPUT_FILE).expect(&format!("Failed to open file '{}'", INPUT_FILE));
    let mut buffer = String::new();
    let mut floor = 0;

    // TODO: we should read byte per byte instead
    file.read_to_string(&mut buffer).expect("read_to_string failed");

    for (index, ch) in buffer.chars().enumerate() {
        match ch {
            '(' => floor += 1,
            ')' => floor -= 1,
            _ => panic!("Invalid character found at position {}", index + 1),
        };

        if floor == -1 {
            println!("Basement is reached at position {}", index + 1);
            process::exit(0);
        }
    }

    println!("The right floor is {}", floor);
}
